app/scripts/modules/core/src/managed/managedResourceDecorators.ts

```diff
@@ -0,0 +1,77 @@
+import { Application } from 'core/application';
+import { SETTINGS } from 'core/config';
+import { IServerGroup, ILoadBalancer, ISecurityGroup } from 'core/domain';
+import { IMoniker } from 'core/naming';
+
+import { IManagedResourceSummary, getResourceKindForLoadBalancerType } from './ManagedReader';
+
+const isMonikerEqual = (a: IMoniker, b: IMoniker) => a.app === b.app && a.stack === b.stack && a.detail === b.detail;
+
+const getResourcesOfKind = (application: Application, kinds: string[]) => {
+  const resources: IManagedResourceSummary[] = application.managedResources.data.resources;
+  return resources.filter(({ kind }) => kinds.includes(kind));
+};
+
+export const addManagedResourceMetadataToServerGroups = (application: Application) => {
+  if (!SETTINGS.feature.managedResources) {
+    return;
+  }
+
+  const clusterResources = getResourcesOfKind(application, ['cluster']);
+  const serverGroups: IServerGroup[] = application.serverGroups.data;
+
+  serverGroups.forEach(serverGroup => {
+    const matchingResource = clusterResources.find(
+      ({ moniker, locations: { account, regions } }) =>
+        isMonikerEqual(moniker, serverGroup.moniker) &&
+        account === serverGroup.account &&
+        regions.some(({ name }) => name === serverGroup.region),
+    );
+
+    serverGroup.isManaged = !!matchingResource;
+    serverGroup.managedResourceSummary = matchingResource;
+  });
+};
+
+export const addManagedResourceMetadataToLoadBalancers = (application: Application) => {
+  if (!SETTINGS.feature.managedResources) {
+    return;
+  }
+
+  const loadBalancerResources = getResourcesOfKind(application, ['classic-load-balancer', 'application-load-balancer']);
```

Comment:
@emjburns @asher @robfletcher @gal-yardeni @luispollo something I started noticing while writing this — the whole thing of having very specific/discrete resource `kind`s is going to gradually build up a bit of an impedance mismatch when it comes to the generic, provider-agnostic modeling we do inside the rest of the API/UI. Here, for example, we have a bunch of load balancers in a list, and I say 'hey application, gimme your load balancers'. But to get the right list of resources that represent all the load balancers which are managed by keel, I need to encode a brittle whitelist of specific resource kinds that I as the programmer know will give me all the currently available load balancer kinds.

Right now LBs are the exception here, because we only have `cluster` and `security-group` otherwise. I'm wondering if we should strive for kinds that are more generic like those, and make load balancer types defined via the spec contents (e.g. `loadBalancerType: 'application'`). There's also a valid (I think) path where we say "no, this is the new world and things are different", but that causes a lot of pretty big and confusing questions to pop up about how we handle cases like this as they become more prevalent. Not proposing we take action on any of this immediately, just raising it as food for thought.

Reply:
I think we should shoot for higher-level abstract "kinds" and add specific details as annotations in the specs. It might be a little trickier to resolve to concrete types under the covers, but to your point I think it's worth the effort because it will not only better align with the UI/API contracts we're exposing, but with the story we're trying to tell with MD. If there's any behavior we want to expose on the UI that is very specific to a concrete type, we could still do that based on the annotations, while keeping the API fairly agnostic.

We kinda talked about that idea here last week as well in the context of Titus vs. other types of container-based deployments: https://github.com/spinnaker/keel/pull/533#discussion_r337232907

Reply:
P.S. On the other hand, I'm trying to be cognizant to avoid "premature optimization" while we don't have a whole lot of customers to provide actual user feedback, so I'm frequently unsure what feedback to give (or hold off) on changing some of these contracts..